pkg/lightning/tikv/prop_collector.go

```diff
@@ -0,0 +1,223 @@
+// Copyright 2024 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tikv
+
+import (
+	"bytes"
+	"encoding/binary"
+	"slices"
+
+	rockssst "github.com/lance6716/pebble/sstable"
+)
+
+type mockCollector struct {
+	name string
+}
+
+var _ rockssst.TablePropertyCollector = mockCollector{}
+
+// Add implements the TablePropertyCollector interface.
+func (mockCollector) Add(rockssst.InternalKey, []byte) error {
+	return nil
+}
+
+// Finish implements the TablePropertyCollector interface.
+func (mockCollector) Finish(map[string]string) error {
+	return nil
+}
+
+// Name implements the TablePropertyCollector interface.
+func (m mockCollector) Name() string {
+	return m.name
+}
+
+type indexHandleKV struct {
+	key    []byte
+	size   uint64
+	offset uint64
+}
+
+// mvccPropCollector is a specialized version of TiKV's `MvccPropertiesCollector`.
+type mvccPropCollector struct {
+	props struct {
+		minTS          uint64
+		maxTS          uint64
+		numRows        uint64
+		numPuts        uint64
+		numDeletes     uint64
+		numVersions    uint64
+		maxRowVersions uint64
+		ttl            struct {
+			maxExpireTS *uint64
+			minExpireTS *uint64
+		}
+	}
+	curIndexSize    uint64
+	curIndexOffset  uint64
+	rowIndexHandles []indexHandleKV
+}
+
+var _ rockssst.TablePropertyCollector = (*mvccPropCollector)(nil)
+
+func newMVCCPropCollector(ts uint64) *mvccPropCollector {
+	ret := &mvccPropCollector{}
+	ret.props.minTS = ts
+	ret.props.maxTS = ts
+	ret.props.maxRowVersions = 1
+	return ret
+}
+
+// Add implements the TablePropertyCollector interface. It mimics
+// https://github.com/tikv/tikv/blob/7793f1d5dc40206fe406ca001be1e0d7f1b83a8f/components/engine_rocks/src/properties.rs#L407.
+func (m *mvccPropCollector) Add(key rockssst.InternalKey, _ []byte) error {
+	m.props.numVersions++
+	m.props.numRows++
+	m.props.numPuts++
+
+	m.curIndexSize++
+	m.curIndexOffset++
+	if m.curIndexOffset == 1 || m.curIndexSize >= 10000 {
```

Comment:
do we need handle for the trailing part

Reply:
https://github.com/pingcap/tidb/pull/57061/commits/87a2a1c7be7bb33008838f6e226c800b8c05beb1

I'll add test about many KVs in next PR, and that should be able to cover this commit